File: src/main/java/com/mkreidl/ephemeris/solarsystem/SolarSystem.java
package com.mkreidl.ephemeris.solarsystem;

import com.mkreidl.ephemeris.Position;
import com.mkreidl.ephemeris.Time;
import com.mkreidl.ephemeris.geometry.Cartesian;
import com.mkreidl.ephemeris.sky.coordinates.Ecliptical;
import com.mkreidl.ephemeris.sky.coordinates.Equatorial;
import com.mkreidl.ephemeris.sky.coordinates.Horizontal;
import com.mkreidl.math.Vector3;

import java.util.EnumMap;
import java.util.Map;

public abstract class SolarSystem {
    
    protected final Map<Body, Ecliptical.Cart> positions = new EnumMap<>(Body.class);
    private final Map<Body, Ecliptical.Cart> velocities = new EnumMap<>(Body.class);
    private final Map<Body, Position> planetsEphemerides = new EnumMap<>(Body.class);
    private final EnumMap<Body, Double> geocentricDistances = new EnumMap<>(Body.class);

    final Map<Body, OrbitalModel> models = new EnumMap<>(Body.class);

    SolarSystem() {
        for (Body body : Body.values()) {
            planetsEphemerides.put(body, new Position());
            positions.put(body, new Ecliptical.Cart());
            velocities.put(body, new Ecliptical.Cart());
        }
    }

    public Cartesian getHeliocentric(final Body body, final Cartesian output) {
        output.set(positions.get(body));
        if (models.get(body).getType() == OrbitalModel.Type.GEOCENTRIC)
            output.add(positions.get(Body.EARTH));
        return output;
    }

    public Position getEphemerides(final Body body, final Position position) {
        switch (models.get(body).getType()) {
            case HELIOCENTRIC:
                position.setHeliocentricPosition(positions.get(body), positions.get(Body.EARTH));
                position.setHeliocentricVelocity(velocities.get(body), velocities.get(Body.EARTH));
                break;
            case GEOCENTRIC:
                position.setGeocentricPosition(positions.get(body), positions.get(Body.EARTH));
                position.setGeocentricVelocity(velocities.get(body), velocities.get(Body.EARTH));
                break;
        }
        position.correctAberration();
        return position;
    }

    public double getGeocentricDistance(Body body) {
        return geocentricDistances.get(body);
    }

    public void compute(final Time time, final Body body) {
        final OrbitalModel model = models.get(body);
        final PhaseCartesian phase = model.computeCartesian(time);
        final Vector3 pos = phase.getPosition().times(model.getDistanceUnit().toMeters());
        final Vector3 vel = phase.getVelocity().times(model.getDistanceUnit().toMeters());
        positions.get(body).set(pos.getX(), pos.getY(), pos.getZ());
        velocities.get(body).set(vel.getX(), vel.getY(), vel.getZ());
    }

    public void compute(final Time time) {
        for (final Body body : Body.values()) {
            compute(time, body);
        }
        final Equatorial.Cart geocentric = new Equatorial.Cart();
        for (final Body body : Body.values()) {
            getEphemerides(body, planetsEphemerides.get(body));
            planetsEphemerides.get(body).get(geocentric, Position.CoordinatesCenter.GEOCENTRIC);
            geocentricDistances.put(body, geocentric.length());
        }
    }

    public void setTimeLocation(final Time time, double longitudeRad, double latitudeRad) {
        final double localSiderealTimeRad = time.getMeanSiderealTimeRadians() + longitudeRad;
        final double currentEclipticRad = new Ecliptic(time.getTime()).getMeanObliquity();
        setTimeLocation(currentEclipticRad, localSiderealTimeRad, latitudeRad);
    }

    public void setTimeLocation(double currentEclipticRadians, double localSiderealTimeRadians, double latitudeRadians) {
        for (final Body body : Body.values())
            planetsEphemerides.get(body).setTimeLocation(currentEclipticRadians, localSiderealTimeRadians, latitudeRadians);
    }

    public void getPosition(Body body, Position.CoordinatesCenter center, Ecliptical.Cart out) {
        planetsEphemerides.get(body).get(out, center);
    }

    public void getPosition(Body body, Position.CoordinatesCenter center, Equatorial.Cart out) {
        planetsEphemerides.get(body).get(out, center);
    }

    public void getPosition(Body body, Position.CoordinatesCenter center, Horizontal.Cart out) {
        planetsEphemerides.get(body).get(out, center);
    }
}

File: src/main/java/com/mkreidl/ephemeris/solarsystem/SolarSystem.kt
package com.mkreidl.ephemeris.solarsystem

import com.mkreidl.ephemeris.Distance
import com.mkreidl.ephemeris.sky.Topos
import com.mkreidl.ephemeris.solarsystem.meeus.*
import com.mkreidl.ephemeris.time.Instant
import com.mkreidl.math.PhaseCartesian
import com.mkreidl.math.PhaseSpherical
import com.mkreidl.math.times

val PhaseCartesian.retrograde get() = position.x * velocity.y - position.y * velocity.x < 0
val PhaseSpherical.retrograde get() = velocity.y < 0

class SolarSystem(private val models: Map<Body, OrbitalModel>) {

    private val eclipticalHeliocentric = mutableMapOf(Body.SUN to PhaseCartesian.ZERO)
    private val eclipticalGeocentric = mutableMapOf(Body.EARTH to PhaseCartesian.ZERO)
    private val geocentricDistances = mutableMapOf(Body.EARTH to 0.0)

    private val toMetersEarth = models.getValue(Body.EARTH).distanceUnit.toMeters()

    private lateinit var ecliptic: Ecliptic

    fun compute(instant: Instant, ecliptic: Ecliptic = Ecliptic(instant)) {
        computeEarth(instant, ecliptic)
        models.keys.forEach { compute(instant, it) }
    }

    fun computeSingle(instant: Instant, ecliptic: Ecliptic = Ecliptic(instant), body: Body) {
        computeEarth(instant, ecliptic)
        compute(instant, body)
    }

    private fun computeEarth(instant: Instant, ecliptic: Ecliptic) {
        this.ecliptic = ecliptic
        eclipticalHeliocentric[Body.EARTH] = models.getValue(Body.EARTH).compute(instant).cartesian * toMetersEarth
        eclipticalGeocentric[Body.SUN] = -correctAberration(eclipticalHeliocentric.getValue(Body.EARTH))
        geocentricDistances[Body.SUN] = eclipticalGeocentric.getValue(Body.SUN).position.norm
    }

    private fun compute(instant: Instant, body: Body) {
        val model = models.getValue(body)
        val earth = eclipticalHeliocentric.getValue(Body.EARTH)
        val toMeters = models.getValue(body).distanceUnit.toMeters()
        if (models.getValue(body).type == OrbitalModel.Type.GEOCENTRIC) {
            val geometricGeocentric = model.compute(instant).cartesian * toMeters
            eclipticalHeliocentric[body] = geometricGeocentric + earth
            eclipticalGeocentric[body] = correctAberration(geometricGeocentric)
        } else {
            val geometricHeliocentric = model.compute(instant).cartesian * toMeters
            eclipticalHeliocentric[body] = geometricHeliocentric
            eclipticalGeocentric[body] = correctAberration(geometricHeliocentric - earth)
        }
        geocentricDistances[body] = eclipticalGeocentric.getValue(body).position.norm
    }

    private fun correctAberration(uncorrected: PhaseCartesian): PhaseCartesian {
        val timeLightTravel = uncorrected.position.norm / Distance.ls.toMeters()
        return uncorrected.copy(position = uncorrected.position - uncorrected.velocity * timeLightTravel)
    }

    fun getGeocentricDistance(body: Body) = geocentricDistances.getValue(body)

    fun getMeanEclipticalHeliocentric(body: Body) = eclipticalHeliocentric.getValue(body)
    fun getMeanEclipticalGeocentric(body: Body) = eclipticalGeocentric.getValue(body)

    fun getTrueEclipticalHeliocentric(body: Body) = ecliptic.trafoMeanEcl2TrueEcl * eclipticalHeliocentric.getValue(body)
    fun getTrueEclipticalGeocentric(body: Body) = ecliptic.trafoMeanEcl2TrueEcl * eclipticalGeocentric.getValue(body)

    fun getMeanEquatorialHeliocentric(body: Body) = ecliptic.trafoEcl2MeanEqu * eclipticalHeliocentric.getValue(body)
    fun getMeanEquatorialGeocentric(body: Body) = ecliptic.trafoEcl2MeanEqu * eclipticalGeocentric.getValue(body)

    fun getTrueEquatorialHeliocentric(body: Body) = ecliptic.trafoMeanEcl2TrueEqu * eclipticalHeliocentric.getValue(body)
    fun getTrueEquatorialGeocentric(body: Body) = ecliptic.trafoMeanEcl2TrueEqu * eclipticalGeocentric.getValue(body)

    fun getMeanEclipticalTopocentric(body: Body, topos: Topos) =
            eclipticalGeocentric.getValue(body).position - topos.meanEcliptical

    fun getTrueEclipticalTopocentric(body: Body, topos: Topos) =
            ecliptic.trafoMeanEcl2TrueEcl * eclipticalGeocentric.getValue(body).position - topos.trueEcliptical

    fun getMeanEquatorialTopocentric(body: Body, topos: Topos) =
            ecliptic.trafoEcl2MeanEqu * eclipticalGeocentric.getValue(body).position - topos.meanEquatorial

    fun getTrueEquatorialTopocentric(body: Body, topos: Topos) =
            ecliptic.trafoMeanEcl2TrueEqu * eclipticalGeocentric.getValue(body).position - topos.trueEquatorial

    /**
     * Compute the fraction of the disk which is lighted by the sun
     */
    fun getIlluminatedFraction(body: Body): Double {
        val helioCart = eclipticalHeliocentric.getValue(body).position
        val geoCart = eclipticalGeocentric.getValue(body).position
        return (1 + (helioCart * geoCart) / (helioCart.norm * geoCart.norm)) * 0.5
    }

    fun getPhase(body: Body) =
            eclipticalHeliocentric.getValue(body).position angle eclipticalGeocentric.getValue(body).position

    /**
     * Compute angle between the rays originating from the earth (observer),
     * pointing to the sun and body, resp.
     *
     * @return Angle in radians
     */
    fun getElongationRadians(body: Body) =
            eclipticalGeocentric.getValue(Body.SUN).position angle eclipticalGeocentric.getValue(body).position

    fun getElongationRadians(body: Body, topos: Topos) =
            getMeanEclipticalTopocentric(Body.SUN, topos) angle getMeanEclipticalTopocentric(body, topos)

    private fun getElongationSign(body: Body) = Math.signum(getElongationRadians(body).radians)

    private fun getElongationSign(body: Body, topos: Topos) = Math.signum(getElongationRadians(body, topos).radians)

    companion object {
        fun createFromMeeus() = SolarSystem(mapOf(
                Body.SUN to SUN,
                Body.MERCURY to MercuryMeeus(),
                Body.VENUS to VenusMeeus(),
                Body.EARTH to EarthMeeus(),
                Body.MARS to MarsMeeus(),
                Body.JUPITER to JupiterMeeus(),
                Body.SATURN to SaturnMeeus(),
                Body.URANUS to UranusMeeus(),
                Body.NEPTUNE to NeptuneMeeus(),
                Body.MOON to ModelMoon(),
                Body.PLUTO to ModelPluto()
        ))

        private object SUN : OrbitalModel {
            override val type = OrbitalModel.Type.HELIOCENTRIC
            override val distanceUnit = Distance.AU
            override val phase = PhaseCartesian.ZERO
            override fun compute(instant: Instant) = phase
        }
    }
}